Refactor environment variable processing for Docker

pugnascotia opened the pull request:
Closes #45223.

The current Docker entrypoint script picks up environment variables and
translates them into -E command line arguments. However, since any tool
executed via `docker exec` doesn't run the entrypoint, it results in
a poorer user experience.

This PR refactors the env var handling so that the -E options are
generated in `elasticsearch-env`. These have to be appended to any
existing command arguments, since some CLI tools have subcommands and
-E arguments must come after the subcommand.

Also extract the support for `_FILE` env vars into a separate script, so
that it can be called from more than once place (the behaviour is
idempotent).

Finally, add noop `-E` handling to `CronEvalTool` for parity, and support `-E` in `MultiCommand` before subcommands.

pugnascotia committed:
Refactor env var processing

Closes #45223.

The current Docker entrypoint script picks up environment variables and
translates them into -E command line arguments. However, since any tool
executes via `docker exec` doesn't run the entrypoint, it results in
a poorer user experience.

Therefore, refactor the env var handling so that the -E options are
generated in `elasticsearch-env`. These have to be appended to any
existing command arguments, since some CLI tools have subcommands and
-E arguments must come after the subcommand.

Also extract the support for `_FILE` env vars into a separate script, so
that it can be called from more than once place (the behaviour is
idempotent).

Finally, change `CronEvalTool` to extend `EnvironmentAwareCommand`, so
that it can allow (but ignore) -E commands. This also brings the tool in
line with our other CLI tools.

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Packaging)

pugnascotia commented on distribution/docker/src/docker/Dockerfile:
```diff
@@ -29,7 +29,7 @@ ${source_elasticsearch}
 RUN tar zxf /opt/${elasticsearch} --strip-components=1
 RUN grep ES_DISTRIBUTION_TYPE=tar /usr/share/elasticsearch/bin/elasticsearch-env \
-    && sed -ie 's/ES_DISTRIBUTION_TYPE=tar/ES_DISTRIBUTION_TYPE=docker/' /usr/share/elasticsearch/bin/elasticsearch-env
+    && sed -i -e 's/ES_DISTRIBUTION_TYPE=tar/ES_DISTRIBUTION_TYPE=docker/' /usr/share/elasticsearch/bin/elasticsearch-env
```
This change is to prevent `sed` creating a backup, which I think was unintentional. I noticed the file `elasticsearch-enve` being created in the Docker image.

dliappis replied:
Great catch. Typically this confusion comes out of the sed version brought by macOS.

dliappis commented on distribution/src/bin/elasticsearch-env:
```diff
@@ -86,4 +86,38 @@ ES_DISTRIBUTION_FLAVOR=${es.distribution.flavor}
+  while IFS='=' read -r envvar_key envvar_value
+  do
+    # Elasticsearch settings need to have at least two dot separated lowercase
+    # words, e.g. `cluster.name`, except for `processors` which we handle
```
Let's remove the comment about `processors` special handling as it's no longer an Elasticsearch setting, see: https://github.com/elastic/elasticsearch/pull/45923

pugnascotia replied:
Removed.

dliappis reviewed (approved):
LGTM! Left a minor comment.

pugnascotia committed:
Remove outdated comment

pugnascotia committed:
Revert CronEvalTool superclass change

Changing `CronEvalTool` to extend `EnvironmentAwareCommand` broke the
unit tests, and after further reflection I decided that this change was
lazy. Actually, the tool should just accept `-E` params but do nothing
with them, instead of pulling in all the reset of the env code.

pugnascotia commented:
So changing `CronEvalTool` to extend `EnvironmentAwareCommand` broke the unit tests, and after further reflection I decided that this change was lazy. Actually, the tool should just accept `-E` params but do nothing with them, instead of pulling in all the reset of the env code. This also means that the tool's help text documents the fact that the `-E options are unused, instead of implying that they do something.

pugnascotia committed:
Fix test after removing elasticsearch-enve

pugnascotia committed:
Allow -E in MultiCommand tools

pugnascotia commented:
I've updated `MultiCommand` to also accept `-E` arguments, since otherwise e.g. `./elasticsearch-node --help` fails when it has `-E args appends. The extra handling does nothing, and does not prevent `-E` arguments later in the command line from being picked up correctly.

pugnascotia commented:
@tvernum I'd appreciate your thoughts on these changes if have capacity.

tvernum commented on libs/cli/src/main/java/org/elasticsearch/cli/MultiCommand.java:
```diff
@@ -45,6 +45,9 @@
         super(description, beforeMain);
+        // Accepting -E here does not prevent subcommands receiving -E arguments, since we also set `posixlyCorrect` below.
+        // This stops option parsing when the first non-option is encountered.
+        parser.accepts("E", "Unused. Pass to a subcommand instead").withRequiredArg();
```
I think this is a problem.
As you know, it's natural for people to expect that passing `-E` as the first argument to a multi-command will behave correctly. Currently they get an error, and have to work out what they've done wrong.

With this change they will no longer get an error, they will just get something that silently ignores their parameter. This is surprising and trappy behaviour.

If we want to do this, then I think we need to either:
1. make it an error to pass `-E` _and_ also call a subcommand (so `--help -E var.from.docker=true` works, but `-E var=true auto` does not); _OR_
2. actually pass the `-E` values into the subcommand like people expect.

tvernum replied:
Unless I'm misunderstanding the behaviour here, and this already does (2), in which case can we make the comment more explicit?

pugnascotia replied:
No, you're quite right. I was thinking primarily of the Docker case, but it would be much better if `-E` was respected wherever it is supplied. I'll look at this again.

pugnascotia replied:
@tvernum I've addressed this in 31e4679 by passing along -E args to subcommands.

pugnascotia committed:
Pass along -E arguments to subcommands

Having MultiCommand silently accept and ignore -E arguments could be
confusing, so instead pass those arguments along to subcommands.

pugnascotia committed:
Checkstyle

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 45223-refactor-env-var-handling

pugnascotia committed:
Restrict new packaging test to default distro

pugnascotia commented:
@elasticmachine run elasticsearch-ci/packaging-matrix

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 45223-refactor-env-var-handling

pugnascotia committed:
Override ES cgroup setting in elasticsearch-env

Although most environment variable handling related to Docker has
already been moved from the entrypoint script to `elasticsearch-env`,
there still remaining the setting of `ES_JAVA_OPTS` to override the
cgroup hierarchy setting. Move this to the env script as well, so that
Docker env var handling is done in one place.

dliappis reviewed (approved):
LGTM for the change in e43283ec1f31cedd81188d62c608f79f28dcc6be

pugnascotia commented:
@jkakavas would you have time to look over the Java parts of this PR from a security POV?

dliappis commented:
@pugnascotia It dawned on me that with this PR we should adjust also the instructions in step 4. pf [configuring-tls-on-docker](https://github.com/elastic/elasticsearch/blob/master/x-pack/docs/en/security/securing-communications/configuring-tls-docker.asciidoc#run-the-example). It runs `elasticsearch-setup-passwords` via `docker exec` so as they are now they explicitly execute the command with `-E ...`.

In a separate PR we can adjust the same instructions in step 3 of the [stack-docs](https://github.com/elastic/stack-docs/blob/master/docs/en/getting-started/get-started-docker.asciidoc#run-in-docker-with-tls-enabled).

tvernum reviewed (approved):
LGTM

pugnascotia committed:
Remove redundant -E flags from docker exec call

pugnascotia commented:
@dliappis I updated the docs in 35487e93187.

dliappis commented on x-pack/docs/en/security/securing-communications/configuring-tls-docker.asciidoc:
```diff
@@ -206,9 +206,6 @@ WARNING: Windows users not running PowerShell will need to remove `\` and join l
 auto --batch \
--Expack.security.http.ssl.certificate=certificates/es01/es01.crt \
--Expack.security.http.ssl.certificate_authorities=certificates/ca/ca.crt \
--Expack.security.http.ssl.key=certificates/es01/es01.key \
```
Thanks! This works, however -- at least this PR -- includes the [UBI work](https://github.com/elastic/elasticsearch/pull/48710) and [earlier in this file](https://github.com/elastic/elasticsearch/pull/49612/files#diff-13bed7d96a74500ddb18bc537d198109L74) we rely on `yum install unzip`. In order to simply things, given that `unzip` is a fairly common requirement for handling certs, what if we include `unzip` in our `Dockerfile` i.e.bundle it in all Elasticsearch images and remove it from the the mentioned line above, so that examples and users don't have to pull it?

pugnascotia replied:
That makes sense to me, and we already include `gzip`.

dliappis replied:
Great. Maybe, esp. given that we have abstracted the package manager invocation now, we could test for the present of `unzip` / `zip` in the path in our test cases, if it's not too complex?

pugnascotia replied:
Good call on the test - I was installing the CLI tools in the wrong part of the Dockefile!

dliappis reviewed (changes requested):
Thanks! While digging into this I realized there are some changes in the images (UBI) which make one docker-compose file in this doc file non functional.

At some point we need docs tests for this, but I left a comment that IMHO we should bundle `unzip` in all our images (given the need to deal with artifacts from cli tools) and simplify the `create-certs.yml` compose file in the docs.

pugnascotia committed:
Add zip / unzip to the Docker image

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 45223-refactor-env-var-handling

dliappis reviewed (approved):
LGTM

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 45223-refactor-env-var-handling

pugnascotia committed:
Check that zip and unzip are installed in Docker

dliappis commented on qa/os/src/test/java/org/elasticsearch/packaging/util/Docker.java:
```diff
@@ -445,6 +448,14 @@ private static void verifyOssInstallation(Installation es) {
+        // These are installed to help users who are working with certificates.
+        Stream.of("zip", "unzip").forEach(cliTool -> {
+            // `which` isn't installed in our image, so instead just try to call the command.
+            final Shell.Result result = dockerShell.runIgnoreExitCode(cliTool + " -h");
```
In bash you can also rely on `type -P <thefile>` if you don't have which to ensure the file is in $PATH:

e.g. on the Ubi image:

```
[elasticsearch@0f6edc7f150a ~]$ type -P zip
[elasticsearch@0f6edc7f150a ~]$ echo $?
1
[elasticsearch@0f6edc7f150a ~]$ type -P ls
/usr/bin/ls
[elasticsearch@0f6edc7f150a ~]$ echo $?
0
```

pugnascotia replied:
`type -P` would work, but as it's a bash builtin, I'd have to run it through bash. That's a actually a bit cumbersome, because we're already running "bash -c 'docker blah blah commmand'". Instead, I'm now using `rpm` to check if the package is installed.

dliappis reviewed (commented):
Great catch! Left a comment

pugnascotia committed:
Use rpm to check for package installation
